tap_snowflake/client.py

```diff
@@ -3,67 +3,228 @@
-        """Return a generator of row-type dictionary objects.
+    def get_batches(
+        self, batch_config: BatchConfig, context: dict | None = None
+    ) -> Iterable[tuple[BaseBatchFileEncoding, list[str]]]:
+        """Get batches of Records from Snowflake.
 
-        Developers may optionally add custom logic before calling the default
-        implementation inherited from the base class.
+        Currently this returns records batches unloaded via an internal user stage.
+        In future this can be updated to include new methods for unloading via external stages.
+        For more details on batch unloading data from Snowflake, see the Snowflake docs:
+        https://docs.snowflake.com/en/user-guide-data-unload.html
+        """
+        if context:
+            raise NotImplementedError(
+                f"Stream '{self.name}' does not support partitioning."
+            )
+        return self.get_batches_from_internal_user_stage(batch_config, context)
 
-        Args:
-            partition: If provided, will read specifically from this data slice.
+    @staticmethod
+    def _get_full_table_copy_statement(
+        sync_id: str, prefix: str, objects: str, table_name: str
+    ) -> Tuple[text, dict]:
+        """Get FULL_TABLE copy statement and key bindings."""
+        return (
+            text(
+                f"copy into '@~/tap-snowflake/{sync_id}/{prefix}' from "
+                + f"(select object_construct({', '.join(objects)}) from {table_name}) "
+                + "file_format = (type='JSON', compression='GZIP') overwrite = TRUE"
+            ),
+            {},
+        )
 
-        Yields:
-            One dict per record.
+    def _get_incremental_copy_statement(
+        self,
+        sync_id: str,
+        prefix: str,
+        objects: str,
+        table_name: str,
+        table,
+        replication_key_value,
+    ) -> Tuple[text, dict]:
+        """Get INCREMENTAL copy statement and key bindings."""
+        return (
+            text(
+                f"copy into '@~/tap-snowflake/{sync_id}/{prefix}' from "
+                + f"(select object_construct({', '.join(objects)}) from {table_name} where {self.replication_key} >= :replication_key_value order by {self.replication_key})"
+                + "file_format = (type='JSON', compression='GZIP') overwrite = TRUE"
+            ),
+            {"replication_key_value": replication_key_value},
+        )
+
+    def _get_copy_statement(
+        self, sync_id: str, prefix: str, context: dict | None = None
+    ) -> Tuple[text, dict]:
+        """Construct copy statement, taking into account stream property selection and incremental keys."""
+        table_name = self.fully_qualified_name
+        table = self.connector.get_table(table_name)
+        selected_columns = self.get_selected_columns(table=table)
+        objects = [f"'{col}', {col}" for col in selected_columns]
+
+        if self.replication_method == REPLICATION_FULL_TABLE:
```

Comment:
@edgarrmondragon @aaronsteers I don't love that this logic has to be handled manually. Apart from being verbose, it also introduces opportunities for errors or mishandling of the bookmark and replication method. It would be great to abstract this away in the SDK by providing methods for users to override for each case:

```python
def get_batches_full_table(self, batch_config: BatchConfig, context: dict | None = None):
    """Replaces `get_batches` for most (i.e. non-incremental) use cases."""
     return NotImplemented

def get_batches_incremental(self, batch_config: BatchConfig, bookmark: Any, context: dict | None = None):
    """Only implemented if filtering is supported.
    Bookmark is supplied, rather than fetched. If not bookmark exists, this method will not be called,
    in favour of `get_batches_full_table`
    """
    return NotImplemented
```

Users can still override `get_batches` if they really want to, but the default implementation does the work to determine a replication strategy before calling `get_batches_full_table` or `get_batches_incremental` respectively.

WDYT?

Reply:
@kgpayne I like it. I worry though that may be too much duplication in implementations of those two methods. Can you implement them in this tap?

Reply:
@kgpayne, @edgarrmondragon - another option is that we just pass something like `lower_bound` and `upper_bound` as nullable. The SDK takes care of sending one or both, and the developer incorporates those limits into the query (or not).

```
def get_batches_incremental(self, batch_config: BatchConfig, bookmark: Any, context: dict | None = None, start_at: Any | None = None, end_at: Any | None = None):
    """
    If `start_at` is defined, records should be filtered to be greater than or equal to this value. (Otherwise unbounded.)
    If `end_at` is defined, records should be filtered to be less than or equal to this value. (Otherwise unbounded.)
    """
    return NotImplemented
```

The same could be implemented for `get_records()` and then devs don't need to call anything at all - they don't need to know what mode they are running.

Reply:
Discussion:

- https://github.com/meltano/sdk/discussions/1011

Reply:
>  I worry though that may be too much duplication in implementations of those two methods. 

Agreed - I prefer the `start_at`/`end_at` suggestion, even if it means changing the signature 👏